### PR TITLE
[compute/cker] Add RmsNorm operation

### PR DESCRIPTION
**Pull request by seockho-kim**
This commit adds RmsNorm operation to cker

ONE-DCO-1.0-Signed-off-by: Seockho Kim seockho.kim@samsung.com

issue: https://github.com/Samsung/ONE/issues/14089
draft: https://github.com/Samsung/ONE/pull/14088

### CODE REVIEW

File: compute/cker/include/cker/operation/RmsNorm.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                    const Shape &gamma_shape, const float *gamma_data, const Shape &beta_shape,
+                    const float *beta_data, const Shape &output_shape, float *output_data)
+{
+  const int32_t batches = MatchingDim(input_shape, 0, output_shape, 0);
+  const int32_t heights = MatchingDim(input_shape, 1, output_shape, 1);
+  const int32_t widths = MatchingDim(input_shape, 2, output_shape, 2);
+  const int32_t channels = MatchingDim(input_shape, 3, output_shape, 3);
+
+  UNUSED_RELEASE(gamma_shape);
+  UNUSED_RELEASE(beta_shape);
```
Comment (ragmani) [inline compute/cker/include/cker/operation/RmsNorm.h:41]: If it won't be used again, could you please remove it from the parameter list?

**Review by ragmani — COMMENTED** [review]
Could you also add unit tests with negative tests for this kernel in "cker/src" directory?

**Comment by seockho-kim** [comment]
> Could you also add unit tests with negative tests for this kernel in "cker/src" directory? We need to satisfy code coverage.

Okay, I'll update it. :)

File: compute/cker/include/cker/operation/RmsNorm.h (hunk 1 above)
Comment (seockho-kim) [reply compute/cker/include/cker/operation/RmsNorm.h:41]: > If it won't be used again, could you please remove it from the parameter list?

Okay, I'll remove it. :)

**Comment by lemmaa** [comment]
@seockho-kim , by the way, as far as I understand, the correct notation for RmsNorm is RMSNorm. How about reflecting the standard name throughout the source code?

**Comment by seockho-kim** [comment]
> @seockho-kim , by the way, as far as I understand, the correct notation for RmsNorm is RMSNorm. How about reflecting the standard name throughout the source code?

Yes, RMSNorm is the correct notation. 
But I just used Pascal case convention in the code.
If "RMSNorm" is better, I'll fix it in another PR. :)

File: compute/cker/include/cker/operation/RmsNorm.h (hunk 1 above)
Comment (seockho-kim) [reply compute/cker/include/cker/operation/RmsNorm.h:41]: > > If it won't be used again, could you please remove it from the parameter list?
> 
> Okay, I'll remove it. :)

I thought it's better to check gamma shape and beta shape in this code.
I'll fix it. :)

**Comment by lemmaa** [comment]
@seockho-kim , if you look at the usage examples of Pascal case or camel case for various programming languages, it seems that there is no one fixed answer.

https://stackoverflow.com/questions/15526107/acronyms-in-camelcase#:~:text=When%20using%20acronyms,%20use%20Pascal%20case%20or%20camel%20case%20for

However, in the case of C++ that we use, I think it is not a bad choice to follow Microsoft's guide.

https://learn.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/141e06ef(v=vs.71)?redirectedfrom=MSDN

image

In conclusion, it would be better to keep your choice. :)

P.S. I think we should also unify the other cases in the ONE source code in the same direction.

**Comment by seockho-kim** [comment]
> @seockho-kim , if you look at the usage examples of Pascal case or camel case for various programming languages, it seems that there is no one fixed answer.
> 
> https://stackoverflow.com/questions/15526107/acronyms-in-camelcase#:~:text=When%20using%20acronyms,%20use%20Pascal%20case%20or%20camel%20case%20for
> 
> However, in the case of C++ that we use, I think it is not a bad choice to follow Microsoft's guide.
> 
> https://learn.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/141e06ef(v=vs.71)?redirectedfrom=MSDN
> 
> image
> 
> In conclusion, it would be better to keep your choice. :)
> 
> P.S. I think we should also unify the other cases in the ONE source code in the same direction.

Thanks for the detailed explanation and reference! :)

**Review by ragmani — APPROVED** [review]
LGTM

File: compute/cker/include/cker/operation/RmsNorm.h [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#include <cmath>
+
+namespace nnfw
+{
+namespace cker
+{
+
+inline void RmsNorm(const RmsNormParams &params, const Shape &input_shape, const float *input_data,
+                    const Shape &gamma_shape, const float *gamma_data, const Shape &beta_shape,
+                    const float *beta_data, const Shape &output_shape, float *output_data)
```
Comment (glistening) [inline compute/cker/include/cker/operation/RmsNorm.h:33]: Is it necessary `beta`? I guess it is `bias`. If so, as I understand, `beta` is not necessary and not used.

- For our purpose model, it does not have `bias`.
- `llama.cpp` and transformer from HF also do not have `beta`.

Comment (seockho-kim) [reply compute/cker/include/cker/operation/RmsNorm.h:33]: > Is it necessary `beta`? I guess it is `bias`. If so, as I understand, `beta` is not necessary and not used.
> 
> * For our purpose model, it does not have `bias`.
> * `llama.cpp` and transformer from HF also do not have `beta`.

Yes, beta(bias) is not necessary for our model, now.
RMSNorm in llama.cpp didn't have beta, too.
But, there is beta(bias) in RMSNorm paper and
I could find 'offset' from the reference code written by RMSNorm author.(Biao Zhang)
https://github.com/bzhangGo/rmsnorm/blob/master/rmsnorm_tensorflow.py

File: compute/cker/include/cker/operation/RmsNorm.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        for (int32_t channel = 0; channel < channels; channel++)
+        {
+          double input_val = input_data[Offset(input_shape, batch, height, width, channel)];
+          square_sum += (input_val * input_val);
+        }
+        double rms = std::sqrt((square_sum / channels) + params.epsilon);
+        for (int32_t channel = 0; channel < channels; channel++)
+        {
+          double gamma = gamma_data[channel];
+          double beta = beta_data[channel];
```
Comment (ragmani) [inline compute/cker/include/cker/operation/RmsNorm.h:63]: if beta data cannot exist, this line should be modified to prevent errors.
For example,
```suggestion
          double beta = 0;
          if (beta_data)
            beta = beta_data[channel];
```

Comment (seockho-kim) [reply compute/cker/include/cker/operation/RmsNorm.h:63]: > if beta data cannot exist, this line should be modified to prevent errors. For example,

Okay, I'll apply it. :)
Gamma may need similar ones.

**Review by ragmani — APPROVED** [review]

File: compute/cker/include/cker/operation/RmsNorm.h (hunk 2 above)
Comment (seockho-kim) [reply compute/cker/include/cker/operation/RmsNorm.h:33]: @glistening 
Please review this.

File: compute/cker/src/RmsNorm.test.cc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+#include <vector>
+
+TEST(CKer_Operation, RmsNorm)
+{
+  // Simple
+  {
+    std::vector<float> input = {0, 1, 2, 3};
+    nnfw::cker::Shape input_shape{1, 2, 2, 1};
+
+    std::vector<float> expected_output = {0, 1, 1, 1};
```
Comment (zetwhite) [inline compute/cker/src/RmsNorm.test.cc:29]: I failed to understand why `expected_output` is `{0, 1, 1, 1}` here. 

I thought that 
 * rms = root(0^2 + 1^2 + 2^2 + 3^2 + e)= root(14 + e) 
 * expected_output = {0, 1/root(14+e), 2/root(14+e), 3/root(14+e)}

Comment (zetwhite) [reply compute/cker/src/RmsNorm.test.cc:29]: Is there sth I missed? Could you help me to understand?

File: compute/cker/src/RmsNorm.test.cc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      EXPECT_NEAR(output[i], expected_output[i], 1e-5f);
+  }
+
+  // Default gamma and beta
+  {
+    std::vector<float> input = {0, 1, 2, 3, 4, 5, 6, 7};
+    nnfw::cker::Shape input_shape{1, 2, 2, 2};
+
+    std::vector<float> expected_output = {0,        1.412802, 0.784404, 1.176606,
+                                          0.883431, 1.104288, 0.920347, 1.073738};
```
Comment (zetwhite) [inline compute/cker/src/RmsNorm.test.cc:55]: Just for curiosity, How did you get these values? 
Did you calculate it manually or using python(pytorch, tensorflow) library?

**Review by zetwhite — APPROVED** [review]
LGTM

File: compute/cker/src/RmsNorm.test.cc (hunk 5 above)
Comment (seockho-kim) [reply compute/cker/src/RmsNorm.test.cc:55]: > Just for curiosity, How did you get these values? Did you calculate it manually or using python(pytorch, tensorflow) library?

I've used pytorch RMSNorm library and wrote RMSNorm function for testing. :)

File: compute/cker/include/cker/operation/RmsNorm.h (hunk 2 above)
Comment (glistening) [reply compute/cker/include/cker/operation/RmsNorm.h:33]: I also checked paper before you wrote comment. I can only find in section 4.1 Invariance Analysis. As I understand, `bias` doesn't matter. re-centering can be ignored for the sake of efficiency w/o sacrificing accuracy much. It is the point of the paper as I understand. It may be the reason the author's code ( personally it does not more important than llama.cpp or our model. It is not the code for production, but for wrting paper ) also disable `bias` by default. Please find:

>     :param bias: whether use bias term for RMSNorm, disabled by
>        default because RMSNorm doesn't enforce re-centering invariance.

Comment (seockho-kim) [reply compute/cker/include/cker/operation/RmsNorm.h:33]: Okay, then I'll remove bias. :)

Comment (ragmani) [reply compute/cker/include/cker/operation/RmsNorm.h]: @ seockho-kim

> Okay, then I'll remove bias. :)

If circle models always don't have `beta(bias)` for RMSNorm op, please remove it from all other code using RMSNorm op.

**Review by ragmani — APPROVED** [review]

**Review by zetwhite — APPROVED** [review]

**Review by glistening — APPROVED** [review]
LGTM

File: compute/cker/include/cker/operation/RmsNorm.h (hunk 2 above)
Comment (seockho-kim) [reply compute/cker/include/cker/operation/RmsNorm.h]: > @seockho-kim
> 
> > Okay, then I'll remove bias. :)
> 
> If circle models always don't have `beta(bias)` for RMSNorm op, please remove the related code from all other code using RMSNorm op.

Sure, I'll remove related codes step by step. :)
